### PR TITLE
Python: model Django HttpRequest as RemoteFlowSource

### PR DESCRIPTION
**Pull request by RasmusWL**
A good chunk of this PR relates to modeling route handlers. This pretty much copies what we do already, so you might want to compare with that: https://github.com/github/codeql/blob/f07a7bf8cff4152845a013fa44001247e796b3a0/python/ql/src/semmle/python/web/django/General.qll

This PR does not try to handle class-based route handlers, I wanted to get started, so let us tackle this in a future PR.

The modeling of taint on the HttpRequest is not complete. I've run out of time for now, and would like to get things moving along, so thought it would be good to get this PR reviewed. Lets extend things further in other PRs.

I realized that current modeling setup for classes and instances (in invoke) does not give us 100% of what we want. I think we're moving closer in this PR, where the `HttpRequest` module has 3 things:

- `classRef()` -- a data-flow node that refers to the `django.http.request.HttpRequest` class (including known aliases)
- `class InstanceSource` -- an abstract class of data-flow nodes that create an instance of `django.http.request.HttpRequest`
- `instance()` -- a data-flow node that refers to an instance of  `django.http.request.HttpRequest`.

The new this is `InstanceSource`. If I want to model a third-party library `cool_lib` that builds on top of django, but has their own logic for setting up routes, the idea is that _in the modeling_ of `cool_lib` I can create a subclass of `InstanceSource` -- this was not possiblein the way `invoke` was modeled. 

The new setup is more like the initial modeling I did for Flask/Werkzeug, that probably needs to be adjusted soonish for consistency. (but I propose we do this in a separate PR).

### CODE REVIEW

**Comment by RasmusWL** [comment]
One thing I did for this PR that I'm not entirely sure was a good idea, was to not mark any inline-expectations with `f-:` when I wanted to handle them but they are currently missing, but instead use the `.expected` file as a TODO list for myself. I can see it has the drawback that when handling a new thing by changing QL code, I'm not removing the `f-:` annotation, so the now correctly handled line of code doesn't show up in the diff.

**Comment by yoff** [comment]
> One thing I did for this PR that I'm not entirely sure was a good idea, was to not mark any inline-expectations with `f-:` when I wanted to handle them but they are currently missing, but instead use the `.expected` file as a TODO list for myself. I can see it has the drawback that when handling a new thing by changing QL code, I'm not removing the `f-:` annotation, so the now correctly handled line of code doesn't show up in the diff.

It also makes it easier to view the PR commit by commit, because the changes are seen next to the line of code and so one does not have to remember what the location of the removed missing result refers to.

**Comment by RasmusWL** [comment]
> It also makes it easier to view the PR commit by commit, because the changes are seen next to the line of code and so one does not have to remember what the location of the removed missing result refers to.

Yep, that's essentially what I tried to write :sweat_smile: I'll try to do better at this moving forwards :+1:

**Review by tausbn — DISMISSED** [review]
Overall I think this looks good. I don't have any objections and/or suggestion for `InstanceSource`. The only thing that gives me pause a bit is the name, as the "source" bit makes me think "taint tracking" rather than type tracking. I have no immediate suggestions for a better name, though.

File: python/ql/test/experimental/library-tests/frameworks/django-v2-v3/routing_test.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+    # one pattern to support `articles/page-<n>` and ensuring that articles/ goes to page-1
+    re_path(r"articles/^(?:page-(?P<page_number>\d+)/)?", show_articles),  # $routeSetup="articles/^(?:page-(?P<page_number>\d+)/)?"
+    # passing as positional argument is not the recommended way of doing things, but it is certainly
+    # possible
+    re_path(r"^([^/]+)/(?:foo|bar)/([^/]+)", xxs_positional_arg, name='xxs_positional_arg'),  # $routeSetup="^([^/]+)/(?:foo|bar)/([^/]+)"
+]
+
+
+# Show we understand the keyword arguments to from django.urls.re_path
```
Comment (yoff) [inline python/ql/test/experimental/library-tests/frameworks/django-v2-v3/routing_test.py:63]:
```suggestion
# Show we understand the keyword arguments to django.urls.re_path
```

**Review by yoff — COMMENTED** [review]
This generally looks great. Only thing that I find slightly puzzling is that you have made your concrete `InstanceSource`, the `HttpRequest`, a `RemoteFlowSource` rather than building one based on the abstract class (so that subclasses in `CoolLib` are automatically included, but perhaps they should not be)?

**Comment by RasmusWL** [comment]
> This generally looks great. Only thing that I find slightly puzzling is that you have made your concrete `InstanceSource`, the `HttpRequest`, a `RemoteFlowSource` rather than building one based on the abstract class (so that subclasses in `CoolLib` are automatically included, but perhaps they should not be)?

That was based on the idea that maybe not all HttpRequest instances are tainted. I'm not aware of any practical implications for this, I just thought it was a "safer" choice (in regards to modeling accuracy) than to mark any future HttpRequest instance as tainted. Maybe I was wrong, I guess time will tell :blush:

**Review by yoff — APPROVED** [review]

**Pull request merged**